### ISSUE TITLE
Multiple (Bidirectional)LSTM layers leads to nan in loss when time step is large

### ISSUE THREAD

**Issue by ScartleRoy** [tensorflow/tensorflow#2995]
I've tried to implement a deep LSTM network and tested it on an autoencoder task in recovering audio magnitude spectrogram. At first I set the number of frames in spectrogram (i.e. time step in LSTM) to be small, and both LSTM and BLSTM worked well. However, when the time step is large, both LSTM and BLSTM began to generate nan loss. To be more specific, when I used a time step of 20 frames, BLSTM worked well with 4 layers and can successfully recover the input; but when I increased it to 100, a 3 layer BLSTM network began to generate nan in two training steps.

I think there might be some problem within the code for `rnn` and `bidirectional_rnn` function (maybe due to unrolling?), but I'm not sure if I did anything wrong.

The codes for (B)LSTM I use are:
```python
def LSTM(lstm_hidden, batch_size, X, chunk_size, name='LSTM', seq_len=None):
    initializer = tf.random_uniform_initializer(-1, 1)
    
    cell = tf.nn.rnn_cell.LSTMCell(lstm_hidden, initializer=initializer, 
                                   use_peepholes=True, state_is_tuple=True)
    
    initial_state = cell.zero_state(batch_size, tf.float32)
    
    if seq_len is not None:     
        output, _ = tf.nn.rnn(cell, X, initial_state=initial_state,
                                             sequence_length=seq_len, scope=name)
    else:
        output, _ = tf.nn.rnn(cell, X, initial_state=initial_state, scope=name)
        
    return output
```
```python
def BLSTM(lstm_hidden, batch_size, X, chunk_size, name='BLSTM', seq_len=None):
    initializer = tf.random_uniform_initializer(-1, 1)
    
    cell_fw = tf.nn.rnn_cell.LSTMCell(lstm_hidden, initializer=initializer, 
                                      use_peepholes=True, state_is_tuple=True)
    cell_bw = tf.nn.rnn_cell.LSTMCell(lstm_hidden, initializer=initializer, 
                                      use_peepholes=True, state_is_tuple=True)

    # initial states
    initial_state_fw = cell_fw.zero_state(batch_size, tf.float32)
    initial_state_bw = cell_bw.zero_state(batch_size, tf.float32)
    
    # BLSTM
    if seq_len is not None:     
        output, _, _ = tf.nn.bidirectional_rnn(cell_fw, cell_bw, X, 
                                               initial_state_fw=initial_state_fw,
                                               initial_state_bw=initial_state_bw, 
                                               sequence_length=seq_len, 
                                               scope=name)
        
    else:
        output, _, _ = tf.nn.bidirectional_rnn(cell_fw, cell_bw, X, 
                                               initial_state_fw=initial_state_fw,
                                               initial_state_bw=initial_state_bw, 
                                               scope=name)
    return output
```

and the loss function is the L2-loss between the recovered magnitude spectrogram and the input, so there might not be a divide-by-zero problem (anyway it works well when time step is small, so I think it's not due to the loss function).

### Environment info
Operating System: Ubuntu 14.04.3
Python version: 2.7.6
Tensorflow version: r0.9
CUDA version: 7.5

**Comment by ScartleRoy** [comment]
@mts42000 Actually I've set the gradient to be 0 for debugging, so this may not be the case. What frustrate me is that the same data and network structure work well in Theano.

**Comment by ScartleRoy** [comment]
Well, finally I found out that the problem occurs because of `initializer = tf.random_uniform_initializer(-1, 1)`. Deleting it will make things begin to work. I saw that initializer [here](https://gist.github.com/kastnerkyle/90a6d0f6789e15c38e97) so I used it, but I still don't know why it doesn't make sense in my case. Anyway just in case someone meets the same problem like I do.

**Comment by gallupliu** [comment]
I would like to know how the multi-layer BLSTM to achieve ？thank you

**Comment by ScartleRoy** [comment]
@gallupliu A simple skeleton is like:
```python
output = input
for n in range(num_layer):
        lstm_fw = LSTMCell(n_hidden, state_is_tuple=True)
        lstm_bw = LSTMCell(n_hidden, state_is_tuple=True)
    
        _initial_state_fw = lstm_fw.zero_state(batch_size, tf.float32)
        _initial_state_bw = lstm_bw.zero_state(batch_size, tf.float32)
    
        output, _states = bidirectional_rnn(lstm_fw, lstm_bw, output, 
                                                  initial_state_fw=_initial_state_fw,
                                                  initial_state_bw=_initial_state_bw, 
                                                  scope='BLSTM_'+str(n+1))
        output_fw = output[0]
        output_bw = output[1]
        output = tf.concat(2, [output_fw, output_bw])
```

Using ```bidirectional_dynamic_rnn``` would make you create the model faster.

**Comment by rajkumarcm** [comment]
@ScartleRoy Could you please explain why you are concatenating the outputs from forward, and backward cells though ?
I thought it should be (as per [Alex Graves, 2013):](http://www.cs.toronto.edu/~graves/asru_2013.pdf)
<pre><code>
fw_out = output[0]
bw_out = output[1]
fw_out = tf.reshape(fw_out,[-1,hidden_size])
bw_out = tf.reshape(bw_out,[-1,hidden_size])
W_fw   = tf.Variable(tf.truncated_normal(shape=[hidden_size,n_chars],
                                         mean=0,stddev=0.1,dtype=tf.float32))
W_bw = tf.Variable(tf.truncated_normal(shape=[hidden_size, n_chars],
                                       mean=0, stddev=0.1, dtype=tf.float32))
b_out = tf.constant(0.1,shape=[n_chars])
logits = tf.add(tf.add(tf.matmul(fw_out,W_fw), tf.matmul(bw_out,W_bw)),b_out)
</code></pre>

**Comment by ScartleRoy** [comment]
@rajkumarcm Oh, that's just the one we're currently in use (and it works). We don't know if the standard one will perform much better than that one, but I think you may choose the standard one in most of the tasks. Well, sorry that I don't know if there are any performance degradation in certain tasks :(

**Comment by rajkumarcm** [comment]
@ScartleRoy Could you please show here how you compute the loss with the concatenated version of logits, and what was the outcome ? 

I have been training a Bidirectional LSTM network, aiming to reproduce result of network described in Alex Grave's "Towards End-End Speech Recognition using Recurrent Neural Networks", but the predicted transcription is nowhere near that of actual. After 4000 iterations (not epochs, mini-batch size:100 with total data: 47,000) I get the CTC loss reducing from 650 to roughly 90. I am finding it really hard to determine the hyperparameters for training such a network.

Any suggestions would be greatly appreciated.

**Comment by ScartleRoy** [comment]
@rajkumarcm I just simply feed the concatenated output into next layer. Sorry that I haven't read that paper (my interest is not in ASR) so that I don't know if the decrease of loss is good or not. But I think you can start with 10% data to see if you can overfit (and determine the size of the network, tune the parameters, etc.). Or maybe choose another learning algorithm (preferably with momentum), start from a large learning rate and gradually decrease it. Those are all very common ways to debug a network.

**Comment by gallupliu** [comment]
@ScartleRoy Thanks，I has master to use the
 function bidirectional_rnn to achieve one layer,But I don‘t know to  achieve multiple layer？may
be I'm wrong understand blstm

**Comment by vino5211** [comment]
@rajkumarcm I am confused about the shape of W_fw you mentioned above: [hidden_size,n_chars],what's n_chars in the second dimension? Is it the max length of the batch of sequences?

**Comment by rajkumarcm** [comment]
@vino5211 Apologies for the late reply. With regards to your question, the second dimension (n_chars) actually refers to n classes ( number of characters + an additional class for blank node ). Hope this helps.
